### PR TITLE
fix: process null values

### PR DESCRIPTION
**Pull request by Kindest13**
Fixes https://github.com/microsoft/kiota-typescript/issues/1188
1 part for processing null values

### CODE REVIEW

File: packages/abstractions/src/serialization/parseNode.ts [hunk 1]
```diff
@@ -31,12 +31,12 @@ export interface ParseNode {
...
 	 * @return the boolean value of the node.
 	 */
-	getBooleanValue(): boolean | undefined;
+	getBooleanValue(): boolean | null | undefined;
 	/**
 	 * Gets the Number value of the node.
 	 * @return the Number value of the node.
 	 */
-	getNumberValue(): number | undefined;
+	getNumberValue(): number | null | undefined;
```
Comment (baywet) [inline packages/abstractions/src/serialization/parseNode.ts:39]: the other types also need to be changed

File: packages/abstractions/src/serialization/serializationWriter.ts [hunk 2]
```diff
@@ -25,19 +25,19 @@ export interface SerializationWriter {
 	 * @param key the key to write the value with.
 	 * @param value the value to write to the stream.
 	 */
-	writeStringValue(key?: string | undefined, value?: string | undefined): void;
+	writeStringValue(key?: string | undefined, value?: string | null | undefined): void;
```
Comment (baywet) [inline packages/abstractions/src/serialization/serializationWriter.ts:28]: same comment about the other types

File: packages/serialization/json/src/jsonParseNode.ts [hunk 3]
```diff
@@ -11,10 +11,10 @@ export class JsonParseNode implements ParseNode {
 	constructor(private readonly _jsonNode: unknown) {}
 	public onBeforeAssignFieldValues: ((value: Parsable) => void) | undefined;
 	public onAfterAssignFieldValues: ((value: Parsable) => void) | undefined;
-	public getStringValue = () => (typeof this._jsonNode === "string" ? (this._jsonNode as string) : undefined);
+	public getStringValue = () => (typeof this._jsonNode === "string" || this._jsonNode == null  ? this._jsonNode : undefined);
```
Comment (baywet) [inline packages/serialization/json/src/jsonParseNode.ts:14]: shouldn't we use triple equals?

File: packages/serialization/json/src/jsonParseNode.ts [hunk 4]
```diff
@@ -11,10 +11,10 @@ export class JsonParseNode implements ParseNode {
...
 	public getChildNode = (identifier: string): ParseNode | undefined => (this._jsonNode && typeof this._jsonNode === "object" && (this._jsonNode as { [key: string]: any })[identifier] !== undefined ? new JsonParseNode((this._jsonNode as { [key: string]: any })[identifier]) : undefined);
-	public getBooleanValue = () => (typeof this._jsonNode === "boolean" ? (this._jsonNode as boolean) : undefined);
-	public getNumberValue = () => (typeof this._jsonNode === "number" ? (this._jsonNode as number) : undefined);
+	public getBooleanValue = () => (typeof this._jsonNode === "boolean" || this._jsonNode == null  ? this._jsonNode : undefined);
```
Comment (baywet) [inline packages/serialization/json/src/jsonParseNode.ts:16]: can you add unit tests for the null case please?

**Review by baywet — CHANGES_REQUESTED** [review]
thanks for starting this. I have left a number of comments to move it forward.

File: packages/abstractions/src/serialization/parseNode.ts (hunk 1 above)
Comment (Kindest13) [reply packages/abstractions/src/serialization/parseNode.ts:39]: changed

File: packages/abstractions/src/serialization/serializationWriter.ts (hunk 2 above)
Comment (Kindest13) [reply packages/abstractions/src/serialization/serializationWriter.ts:28]: changed

File: packages/serialization/json/src/jsonParseNode.ts (hunk 3 above)
Comment (Kindest13) [reply packages/serialization/json/src/jsonParseNode.ts]: agree, fixed

File: packages/serialization/json/src/jsonParseNode.ts (hunk 4 above)
Comment (Kindest13) [reply packages/serialization/json/src/jsonParseNode.ts]: added

File: packages/serialization/json/src/jsonSerializationWriter.ts [hunk 5]
```diff
@@ -75,6 +110,11 @@ export class JsonSerializationWriter implements SerializationWriter {
 		}
 	};
 	public writeCollectionOfObjectValues = <T extends Parsable>(key: string, values: T[], serializerMethod: ModelSerializerFunction<T>): void => {
```
Comment (baywet) [inline packages/serialization/json/src/jsonSerializationWriter.ts:112]:
```suggestion
	public writeCollectionOfObjectValues = <T extends Parsable>(key: string, values: T[] | null | undefined, serializerMethod: ModelSerializerFunction<T>): void => {
```

(and repeat for other object/enum methods)

File: packages/serialization/json/src/jsonSerializationWriter.ts [hunk 6]
```diff
@@ -214,7 +254,7 @@ export class JsonSerializationWriter implements SerializationWriter {
 		}
 		this.writer.push(JSON.stringify(value), JsonSerializationWriter.propertySeparator);
 	};
-	private readonly writeAnyValue = (key?: string | undefined, value?: unknown | undefined): void => {
+	private readonly writeAnyValue = (key?: string | undefined, value?: unknown | null | undefined): void => {
```
Comment (baywet) [inline packages/serialization/json/src/jsonSerializationWriter.ts:257]: the conditions should be reviewed to write null only when the value is null, and not when it's undefined

Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts:257]: should I write `undefined` when it's `undefined` or do nothing?

Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts:257]: @baywet

File: packages/serialization/json/src/jsonSerializationWriter.ts (hunk 5 above)
Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts:112]: should I touch other places? `form`, `text`
image

File: packages/serialization/json/src/jsonParseNode.ts [hunk 7]
```diff
@@ -11,15 +11,21 @@ export class JsonParseNode implements ParseNode {
...
-	public getStringValue = () => (typeof this._jsonNode === "string" ? (this._jsonNode as string) : undefined);
-	public getChildNode = (identifier: string): ParseNode | undefined => (this._jsonNode && typeof this._jsonNode === "object" && (this._jsonNode as { [key: string]: any })[identifier] !== undefined ? new JsonParseNode((this._jsonNode as { [key: string]: any })[identifier]) : undefined);
-	public getBooleanValue = () => (typeof this._jsonNode === "boolean" ? (this._jsonNode as boolean) : undefined);
-	public getNumberValue = () => (typeof this._jsonNode === "number" ? (this._jsonNode as number) : undefined);
-	public getGuidValue = () => parseGuidString(this.getStringValue());
-	public getDateValue = () => (this._jsonNode ? new Date(this._jsonNode as string) : undefined);
-	public getDateOnlyValue = () => DateOnly.parse(this.getStringValue());
-	public getTimeOnlyValue = () => TimeOnly.parse(this.getStringValue());
-	public getDurationValue = () => Duration.parse(this.getStringValue());
+	public getStringValue = () => (typeof this._jsonNode === "string" || this._jsonNode === null ? this._jsonNode : undefined);
```
[the remark this answers is not in the document]
Comment (Kindest13) [reply packages/serialization/json/src/jsonParseNode.ts:14]: basically I thought that this is the place where we do parsing of our json body so that's why added functionality to allow `null` values.
Is it the wrong assumption?

File: packages/serialization/json/src/jsonSerializationWriter.ts (hunk 5 above)
Comment (baywet) [reply packages/serialization/json/src/jsonSerializationWriter.ts:112]: Yes things need to be aligned between the interface season and the implementations.

File: packages/serialization/json/src/jsonParseNode.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (baywet) [reply packages/serialization/json/src/jsonParseNode.ts:14]: Let's try to break this down. The properties of a model need to have null in the type definition so the user can set a null value. This also means that the argument of serialization writer need to accept null as a value so we can properly write it to the serialized representation.
Then we have mostly two choices:
- If the return type of parse node methods does not include null as a valid type it's going to be ok because it's still going to be a subset of what the property we're trying to assign accepts. During the deserialization we can treat null as functionally equivalent to undefined.
- If the return type of the parsnip methods does include null as a valid type then we need to make a difference between the serialization and return null.

The question then becomes do users care about telling the difference between null and undefined when it is returned by the service they are talking to? I would argue they do not since we've never had this request for any other language at this point. I would like to get you and @andrueastman opinion on the matter before we make a final decision on this.

[the remark this answers is not in the document]
Comment (andrueastman) [reply packages/serialization/json/src/jsonParseNode.ts:14]: > we've never had this request for any other language at this point.
Yeah. I would look at it through this lens. 
It would not be best placed to conceive of requirements here without a solid backing. The working default state as `undefined` would be okay for now and we could always change this once we get solid evidence of the need for distinguishing null and undefined for service level responses.

[the remark this answers is not in the document]
Comment (baywet) [reply packages/serialization/json/src/jsonParseNode.ts:14]: Thanks for chiming in here Andrew. 
@kindest13 are you clear on the modifications to be done at this point or do you have any additional questions?

[the remark this answers is not in the document]
Comment (Kindest13) [reply packages/serialization/json/src/jsonParseNode.ts:14]: As I understood there's no need to check on `null` value so just adding the `null` type is sufficient.
Although if `getStringValue` received `null` it returns `undefined` then how do we differenciate between `null` and `undefined`?

File: packages/serialization/json/test/common/testEntity.ts [hunk 8]
```diff
@@ -11,17 +11,17 @@ import { Guid } from "guid-typescript";
...
 	additionalData?: Record<string, unknown>;
-	testDate?: Date | undefined;
+	testDate?: Date | null | undefined;
 	foos?: FooResponse[] | undefined;
-	id?: string | undefined;
-	testNumber?: number | undefined;
-	testGuid?: Guid | undefined;
+	id?: string | null | undefined;
+	testNumber?: number | null | undefined;
+	testGuid?: Guid | null | undefined;
```
[the remark this answers is not in the document]
Comment (Kindest13) [reply packages/serialization/json/test/common/testEntity.ts:24]: which places are we talking about?

File: packages/serialization/json/src/jsonParseNode.ts [hunk 9]
```diff
@@ -53,7 +59,11 @@ export class JsonParseNode implements ParseNode {
 		}
 		return undefined;
 	}
-	public getCollectionOfObjectValues = <T extends Parsable>(method: ParsableFactory<T>): T[] | undefined => {
+	public getCollectionOfObjectValues = <T extends Parsable>(method: ParsableFactory<T>): T[] | null | undefined => {
+		if (this._jsonNode === null) {
+			return null;
+		}
+
```
Comment (Kindest13) [inline packages/serialization/json/src/jsonParseNode.ts:66]: what about these cases?
should I also remove `null` handling?
Will they handle it properly?

File: packages/serialization/json/src/jsonParseNode.ts [hunk 10]
```diff
@@ -11,15 +11,21 @@ export class JsonParseNode implements ParseNode {
...
+	public getDateValue = () => {
+		if (this._jsonNode === null) {
+			return null;
+		}
+
+		return this._jsonNode ? new Date(this._jsonNode as string) : undefined;
+	};
+	public getDateOnlyValue = () => (this.getStringValue() === null ? null : DateOnly.parse(this.getStringValue()));
+	public getTimeOnlyValue = () => (this.getStringValue() === null ? null : TimeOnly.parse(this.getStringValue()));
+	public getDurationValue = () => (this.getStringValue() === null ? null : Duration.parse(this.getStringValue()));
```
Comment (Kindest13) [inline packages/serialization/json/src/jsonParseNode.ts:28]: @baywet what about these cases?
should I also remove null handling?
Will they handle it properly?

File: packages/serialization/json/src/jsonParseNode.ts [hunk 11]
```diff
@@ -53,7 +59,11 @@ export class JsonParseNode implements ParseNode {
 		}
 		return undefined;
 	}
-	public getCollectionOfObjectValues = <T extends Parsable>(method: ParsableFactory<T>): T[] | undefined => {
+	public getCollectionOfObjectValues = <T extends Parsable>(method: ParsableFactory<T>): T[] | null | undefined => {
+		if (this._jsonNode === null) {
+			return null;
```
Comment (Kindest13) [inline packages/serialization/json/src/jsonParseNode.ts:64]: @baywet  -//- https://github.com/microsoft/kiota-typescript/pull/1248/files#r1664428197

File: packages/serialization/json/src/jsonParseNode.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (baywet) [reply packages/serialization/json/src/jsonParseNode.ts:14]: Let me reiterate so we make sure we're all on the same page. First, the models needs to be generated with the type, undefined, null. This is so people can assign null to properties. Then, the serialization writer method parameters need to be able to accept null as well. This is so we can write the new value to the serialized representation. Lastly, the parse node interface does not need to be updated but the implementation needs to treat no from a serialized value as it treats undefined today.
Does this clarify the changes that need to be made?

File: packages/serialization/json/src/jsonParseNode.ts (hunk 10 above)
Comment (baywet) [reply packages/serialization/json/src/jsonParseNode.ts:28]: If we encounter null in a serialized response we need to treat it as undefined and return undefined instead.

File: packages/serialization/json/src/jsonParseNode.ts (hunk 11 above)
Comment (baywet) [reply packages/serialization/json/src/jsonParseNode.ts:64]: same response

File: packages/serialization/json/test/common/testEntity.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kindest13) [reply packages/serialization/json/test/common/testEntity.ts:24]: @baywet  ?

File: packages/serialization/json/src/jsonSerializationWriter.ts (hunk 5 above)
Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts:112]: @baywet last on this, should I care about `multipart` serialization?
image

Comment (baywet) [reply packages/serialization/json/src/jsonSerializationWriter.ts]: meaning: should you write a null value? or did you ask a different question?

Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts]: yes and add `null` type

Comment (baywet) [reply packages/serialization/json/src/jsonSerializationWriter.ts]: yes, I think we should both have null in the parameter type, and write null to json when the value is null (but not when it's undefined)

File: packages/serialization/json/test/common/testEntity.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (baywet) [reply packages/serialization/json/test/common/testEntity.ts:24]: form and multipart tests also have a test entity

File: packages/serialization/json/src/jsonSerializationWriter.ts (hunk 5 above)
Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts]: @baywet Should I touch the `writeByteArrayValue`?

image

Comment (Kindest13) [reply packages/serialization/json/src/jsonSerializationWriter.ts]: @baywet I think I did it, please review
image

File: packages/abstractions/src/serialization/kiotaJsonSerializer.ts [hunk 12]
```diff
@@ -56,7 +56,7 @@ export function serializeCollectionToJsonAsString<T extends Parsable>(values: T[
  * @param factory the factory for the model type
  * @returns the deserialized parsable object
  */
-export function deserializeFromJson<T extends Parsable>(bufferOrString: ArrayBuffer | string, factory: ParsableFactory<T>): Parsable {
+export function deserializeFromJson<T extends Parsable>(bufferOrString: ArrayBuffer | string, factory: ParsableFactory<T>): Parsable | null {
```
Comment (baywet) [inline packages/abstractions/src/serialization/kiotaJsonSerializer.ts:59]: I believe this file needs to be reverted

File: packages/abstractions/src/serialization/kiotaSerializer.ts [hunk 13]
```diff
@@ -85,7 +85,7 @@ function getStringValueFromBuffer(buffer: ArrayBuffer): string {
  * @param factory the factory for the model type
  * @returns the deserialized parsable object
  */
-export function deserialize<T extends Parsable>(contentType: string, bufferOrString: ArrayBuffer | string, factory: ParsableFactory<T>): Parsable {
+export function deserialize<T extends Parsable>(contentType: string, bufferOrString: ArrayBuffer | string, factory: ParsableFactory<T>): Parsable | null {
```
Comment (baywet) [inline packages/abstractions/src/serialization/kiotaSerializer.ts:88]: I believe this file needs to be reverted

File: packages/abstractions/src/serialization/parseNode.ts [hunk 14]
```diff
@@ -20,75 +20,75 @@ export interface ParseNode {
 	 * Gets the string value of the node.
 	 * @return the string value of the node.
 	 */
-	getStringValue(): string | undefined;
+	getStringValue(): string | null | undefined;
```
Comment (baywet) [inline packages/abstractions/src/serialization/parseNode.ts:23]: given our latest discussions, I believe this file needs to be reverted.

File: packages/serialization/form/src/formParseNode.ts [hunk 15]
```diff
@@ -36,7 +36,7 @@ export class FormParseNode implements ParseNode {
 	public onBeforeAssignFieldValues: ((value: Parsable) => void) | undefined;
 	public onAfterAssignFieldValues: ((value: Parsable) => void) | undefined;
 	public getStringValue = (): string => decodeURIComponent(this._rawString);
-	public getChildNode = (identifier: string): ParseNode | undefined => {
+	public getChildNode = (identifier: string): ParseNode | null | undefined => {
```
Comment (baywet) [inline packages/serialization/form/src/formParseNode.ts:39]: this file needs to be reverted as well

File: packages/serialization/json/src/jsonParseNode.ts [hunk 16]
```diff
@@ -12,15 +12,15 @@ export class JsonParseNode implements ParseNode {
 	public onBeforeAssignFieldValues: ((value: Parsable) => void) | undefined;
 	public onAfterAssignFieldValues: ((value: Parsable) => void) | undefined;
```
Comment (baywet) [inline packages/serialization/json/src/jsonParseNode.ts:13]: the prototype changes on this file need to be reverted

File: packages/serialization/json/test/common/JsonParseNode.ts [hunk 17]
```diff
@@ -70,16 +70,16 @@ describe("JsonParseNode", () => {
 		assert.equal(enumValueResult, TestEnumObject.A);
 	});
 
-	it("Test a null collection of object values", async () => {
+	it("Test an undefined collection of object values", async () => {
```
Comment (baywet) [inline packages/serialization/json/test/common/JsonParseNode.ts:73]: we also need a test for null

File: packages/serialization/text/src/textParseNode.ts [hunk 18]
```diff
@@ -17,7 +17,7 @@ export class TextParseNode implements ParseNode {
 			this.text = this.text.substring(1, this.text.length - 2);
 		}
 	}
-	public getByteArrayValue(): ArrayBuffer | undefined {
+	public getByteArrayValue(): ArrayBuffer | null | undefined {
```
Comment (baywet) [inline packages/serialization/text/src/textParseNode.ts:20]: this file needs to be reverted

File: packages/serialization/text/src/textSerializationWriter.ts [hunk 19]
```diff
@@ -19,53 +19,77 @@ export class TextSerializationWriter implements SerializationWriter {
 		this.writeStringValue(key, b64);
 	}
 	private static noStructuredDataMessage = "text does not support structured data";
-	private readonly writer: string[] = [];
+	private readonly writer: (string | null)[] = [];
```
Comment (baywet) [inline packages/serialization/text/src/textSerializationWriter.ts:22]: why not simply write the null text representation and keep it as an array of strings?

**Review by baywet — CHANGES_REQUESTED** [review]
I think we're making good progress here! Thanks for the continuous efforts.

File: packages/serialization/json/test/common/JsonParseNode.ts (hunk 17 above)
Comment (Kindest13) [reply packages/serialization/json/test/common/JsonParseNode.ts:73]: @baywet `null`, also returns `undefined` , right?

Comment (Kindest13) [reply packages/serialization/json/test/common/JsonParseNode.ts:73]: image

Comment (baywet) [reply packages/serialization/json/test/common/JsonParseNode.ts:73]: > @baywet `null`, also returns `undefined` , right?

correct.

I'm not sure I understand the second comment though?

**Review by baywet — COMMENTED** [review]
I've just pushed a set of changes to fix the unit test for form serialization writer. I think similar fixes are required for the json serialization writer as its test is also failing.

**Comment by Kindest13** [comment]
> I've just pushed a set of changes to fix the unit test for form serialization writer. I think similar fixes are required for the json serialization writer as its test is also failing.

@baywet added

File: packages/serialization/json/test/common/JsonParseNode.ts (hunk 17 above)
Comment (Kindest13) [reply packages/serialization/json/test/common/JsonParseNode.ts:73]: nvm, it's just the screenshot of `null` case test

**Review by baywet — DISMISSED** [review]
LGTM!
@andrueastman for a final review

**Comment by Kindest13** [comment]
> LGTM! @andrueastman for a final review

Great! I just comitted changes to remove `.only` from test file

**Review by baywet — DISMISSED** [review]

**Review by andrueastman — APPROVED** [review]
LGTM.

@Kindest13 I believe you still need to reply to the policy service bot to get this merged.

**Comment by rners01** [comment]
> @Kindest13 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by Kindest13** [comment]
> @Kindest13 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by baywet** [comment]
Now this PR is ready, I'd like us to get started on the generation updates before we merge this pull request.
@Kindest13 do you have all the information you need to get started in the kiota repository or do you want additional pointers?

**Comment by Kindest13** [comment]
> Now this PR is ready, I'd like us to get started on the generation updates before we merge this pull request. @Kindest13 do you have all the information you need to get started in the kiota repository or do you want additional pointers?

Would be great if you can share more info on this

**Comment by baywet** [comment]
My guess is that since the `?` is hardcoded here. And since only the properties need to be settable to null, we could probably hardcoded it here.
https://github.com/microsoft/kiota/blob/4d59461de79d1de58a309b4c25fee7b005a15ee5/src/Kiota.Builder/Writers/TypeScript/CodePropertyWriter.cs#L37

Let us know if you have additional questions.

**Comment by baywet** [comment]
` | null` roughly

**Comment by rners01** [comment]
> hardcoded

Hey @baywet, is that correct?
So just this line?
`writer.WriteLine($"{codeElement.Name.ToFirstCharacterLowerCase()}?: {returnType}{(isFlagEnum ? "[]" : string.Empty)} | null;");`

**Comment by Kindest13** [comment]
> ` | null` roughly

Hey @baywet, is that correct?
So just this line?
`writer.WriteLine($"{codeElement.Name.ToFirstCharacterLowerCase()}?: {returnType}{(isFlagEnum ? "[]" : string.Empty)} | null;");`

**Comment by baywet** [comment]
yes it looks correct, however I left a couple of comments on the PR to wrap it up.

**Review by baywet — APPROVED** [review]
